lxd/api.go

```diff
@@ -208,12 +214,98 @@ func restServer(d *Daemon) *http.Server {
 		_ = response.NotFound(nil).Render(w)
 	})
 
+	// Initialize API metrics and use middleware functions to update API request metrics.
+	metrics.InitAPIMetrics()
+	mux.Use(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Set the callback function to track the request as completed.
+			// Assure it can be called at most once.
+			var once sync.Once
+			callbackFunc := func(result metrics.RequestResult) {
+				once.Do(func() {
+					metrics.TrackCompletedRequest(*r.URL, result)
```

Comment:
I'm thinking we should store the original request URL into a local var in this handler function (not the once.Do func) and use it both here and i metrics.TrackStartedRequest below, so that if the r.URL is changed between the request starting and finishing we don't end up losing track of the result.